### ISSUE TITLE
Git hooks are not triggered on commit (flatpak)

### ISSUE THREAD

**Issue by alex-ball** [Murmele/Gittyup#261]
I am running Gittyup v1.1.2 on Ubuntu 20.04 via flatpak.

**Expected behaviour:**
In a Git repo with executable files at `.git/hooks/pre-commit` and `.git/hooks/post-commit`, committing some changes should run those two scripts immediately before and after the commit, respectively.

**Actual behaviour:**
The commit succeeds but the two files do not execute.

I'm afraid I don't know enough about these things to know if

1. this is a security limitation of flatpaks and can only be solved by installing some other way;
2. this can be solved by editing the app permissions;
3. this is a bug in the code;
4. this would be an enhancement.

I have not tested whether the other Git hooks are triggered (by push, receive, etc.).

**Comment by exactly-one-kas** [comment]
Currently, only the `pre-push` hook is supported, although I'm not sure if this works with flatpak.
@Murmele do you know if flatpak needs additional permissions to execute programs outside its sandbox even if they're within opened directories?

**Comment by exactly-one-kas** [comment]
Also, the `pre-push` requires bash for some reason? I think this restriction should be lifted

**Comment by Murmele** [comment]
currently for the merge commands we are leaving the flatpak sandbox. 
there we are using `flatpak-spawn --host sh -c "<command>"` to execute a command on the host system. It works withouth any additional permissions

**Comment by Murmele** [comment]
It depends on what is written in the script. If some host commands are used then it will not work. If just common commands are used I have to check

**Comment by Murmele** [comment]
@exactly-one-kas how it behaves without sandbox? Do I have to do chmod +x first on those files, or is git able to execute them without?

**Comment by exactly-one-kas** [comment]
Git calls them as executables and requires the executable bit to be set, so I'd say we should use `flatpack-spawn --host "<path to hook>"`

**Comment by Murmele** [comment]
> Git calls them as executables and requires the executable bit to be set, so I'd say we should use `flatpack-spawn --host "<path to hook>"`

yes I think that makes it easier. The question is are we then going around some sandboxing features?

**Comment by exactly-one-kas** [comment]
Since we're running arbitrary scripts on the host, I'd say yes. Can we run them inside the sandbox or is that prevented?

**Comment by Murmele** [comment]
> 
I think that should be also possible.

**Comment by schnedann** [comment]
I just implemented a prepare-commit-hook and learned gittyup does not execute it. I see this functionality as very important, as the hooks are one of the few ways to mitigate some of the worst design fails of git...

So from me: please, hooks are important, please implement

**Comment by anselme16** [comment]
Hi, so if i understand well, git hooks are not supported yet only in the flatpak version ? If i compile it from source i will not be sandboxed and the pre-commit hook will be executed ?

**Comment by Murmele** [comment]
No, currently we do not support hooks. The problem is that libgit2 does not
support this. In this case we have to execute them manually and this is not
yet implemented

Anselme FRANÇOIS ***@***.***> schrieb am Mo., 4. Nov. 2024,
13:35:

**Comment by Murmele** [comment]
> Thanks.
> 
> By the way for context i found the related libgit2 issue where they discuss on hook support : [libgit2/libgit2#964](https://github.com/libgit2/libgit2/issues/964)
> 
> They will indeed never support hooks because they don't want to be responsible for calling external processes, so they assume every app using libgit2 will have to reimplement their own hooks.

Yes exactly

**Comment by Murmele** [comment]
We have already a few external calls, for example for the merge tool / diff tool, so this code could be checked out for implementation details

https://github.com/Murmele/Gittyup/blob/80231ac219d45edc579dd4d65dc255861ec585ff/src/tools/EditTool.cpp#L75-L91

**Comment by schnedann** [comment]
Of course its unsafe - but it is git which introduced the hooks, so a lib which mimics git command.... anyway, one could improve security if the app at first start scans the hooks and stores checksums for all hook files. Then, if there is a new or changed file, request the users consent...
Thats not really safe, also but if one would be able to commit the file containing the checksums, chances are good an altered file would be detected.

Or what about use a VCS for the hook files? :-)

**Comment by Murmele** [comment]
We could have already that problem with the merge tools. Just fake for example kdiff3 by adding a script called the name of the mergetool somewhere and if the path is before the correct binary, we execute that script instead of the real merge tool :)
In this case we could bundle all merge tools within the package, but it is a lot more effort

**Comment by amirreza8002** [comment]
any updates on this?
pre-commit is used reqularly in my projects

and my CI breaks if the hooks don't run :)

**Comment by rhvonlehe** [comment]
I'll chime in and say this is a feature I'm really wanting, too.
